teuthology/provision/maas.py

```diff
@@ -0,0 +1,441 @@
+        
+        :returns: A requests output
+        """
+        args = {"url": config.maas["api_url"] + path}
+        args["data"] = json.dumps(data) if data else None
+        args["params"] = params if params else None
+        args["files"] = files if files else None
+
+        resp = None
+        if method == "GET":
+            resp = self.session.get(**args)
+        elif method == "POST":
+            resp = self.session.post(**args)
+        elif method == "PUT":
+            resp = self.session.put(**args)
+        elif method == "DELETE":
+            resp = self.session.delete(**args)
+        else:
+            raise RuntimeError("Unsupported HTTP method '%s'", method)
+
+        if not resp.ok:
+            self.log.error(
+                "Got status %s from %s: '%s'",
+                resp.status_code, path, resp.text
+            )
+
+        if raise_on_error:
+            resp.raise_for_status()
+
+        return resp
+
+    def get_machines_data(self):
+        """Locate the machine we want to use
+
+        returns: The machine data as a dict
+        """
+        resp = self.do_request(
+            "/machines/", params={"hostname": self.shortname}
+        ).json()
+        if len(resp) == 0:
+            raise RuntimeError("Machine %s not found!" % self.shortname)
+        if len(resp) > 1:
+            raise RuntimeError(
+                "More than one machine found for %s" % self.shortname
+            )
+        return resp[0]
+
+    def get_image_data(self):
+        """Locate the image we want to use
+
+        :returns: The image data as a dict
+        """
+        resp = self.do_request("/boot-resources/").json()
+        if len(resp) == 0:
+            raise RuntimeError("MaaS has no images available.")
+
+        name = "%s/%s" % (
+            self.os_type,
+            ubuntu_codenames.get(self.os_version)
+            if self.os_type == "ubuntu" else self.os_version
+        )
+        for image in resp:
+            if image["name"] == name:
+                return image
+        raise RuntimeError("MaaS has no %s image." % name)
+
+    def lock_machine(self, raise_on_error=True):
+        """Lock the machine
+
+        :param raise_on_error: Whether or not to raise an exception
+                               if the request is unsuccessful (default: True)
+        """
+        resp = self.do_request(
+            f"/machines/{self.system_id}/op-lock",
+            method="POST",
+            raise_on_error=raise_on_error
+        )
+        if resp.json():
+            if not resp.json().get("locked") and raise_on_error:
+                raise RuntimeError(
+                    "Machine %s locking failed, status: %s",
+                    self.shortname, resp.json().get("locked")
+                )
+        elif raise_on_error:
+            raise RuntimeError(
+                "Failed to lock machine %s, system_id: %s",
+                self.shortname, self.system_id
```

Comment:
Is there anything else we can tell the user that might help them understand _why_ the failure occurred? I have the same question about other error messages in this file.